### ISSUE TITLE
NPE in DefaultRecordStore on remove

### ISSUE THREAD

**Issue by ivenhov** [hazelcast/hazelcast#15217]
Hi

I have NPE using Hazelcast 3.9.5 which I believe will apply also to 3.12
I'm doing conditional remove with key of type String and value String via

map.remove("keyA", "valA")

Map configuration as below

```
    <map name="linkToMap">
        <backup-count>2</backup-count>
        <eviction-policy>LRU</eviction-policy>
        <max-size policy="PER_NODE">100000</max-size>
        <near-cache>
            <time-to-live-seconds>0</time-to-live-seconds>
            <max-idle-seconds>6000</max-idle-seconds>
            <eviction eviction-policy="LRU" max-size-policy="ENTRY_COUNT" size="5000"/>
            <invalidate-on-change>true</invalidate-on-change>
        </near-cache>
        <map-store enabled="true" initial-mode="LAZY">
            <class-name>x.y.z.DDBBridgeLinkTo</class-name>
            <write-delay-seconds>0</write-delay-seconds>
        </map-store>
    </map>
```

Value is deleted from backend but NPE is thrown as below

```
ERROR 2019-06-25 10:55:44.271  [x.partition-operation.thread-1] log()                          [10.173.240.1]:7008 [matrix] [3.9.4] null      (RemoveIfSameOperation.java:69)
java.lang.NullPointerException: null
        at com.hazelcast.map.impl.recordstore.DefaultRecordStore.remove(DefaultRecordStore.java:609)
        at com.hazelcast.map.impl.operation.RemoveIfSameOperation.run(RemoveIfSameOperation.java:41)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:194)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:120)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
WARN  2019-06-25 10:55:44.273  [DeletionThread           ] doRun()                        Exception during deletion of 27232ed1-9722-11e9-8657-998555bccbfe. Will be retried later      (LocalDeletionService.java:151)
java.lang.NullPointerException: null
        at com.hazelcast.map.impl.recordstore.DefaultRecordStore.remove(DefaultRecordStore.java:609)
        at com.hazelcast.map.impl.operation.RemoveIfSameOperation.run(RemoveIfSameOperation.java:41)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:194)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:120)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
        at ------ submitted from ------.(Unknown Source)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:79)
        at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
        at com.hazelcast.map.impl.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:424)
        at com.hazelcast.map.impl.proxy.MapProxySupport.removeInternal(MapProxySupport.java:575)
        at com.hazelcast.map.impl.proxy.NearCachedMapProxyImpl.removeInternal(NearCachedMapProxyImpl.java:337)
        at com.hazelcast.map.impl.proxy.MapProxyImpl.remove(MapProxyImpl.java:216)
```

This is caused by "record" variable being null.
In v3.9.5
https://github.com/hazelcast/hazelcast/blob/3.9.5/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

I believe the same thing will happen on 3.12 but on line 499
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

Is this a known issue or am I doing something wrong?

**Comment by ahmetmircik** [comment]
Hi, thanks for reporting this issue. It is not a known one. Target release for fix is 3.12.2.

**Comment by ivenhov** [comment]
Thanks for prompt reply.
Any chance it will be backported to 3.9 ?

Regards
Daniel

**Comment by mmedenjak** [comment]
Hi @ivenhov ! Unfortunately we only provide patch releases only for the latest released minor version, which is now 3.12. So we will not backport it to 3.9. But I always recommend moving to newer releases as they always have more fixes, more features and are generally more stable.

**Comment by ivenhov** [comment]
Thanks Matko.
We've moved recently to 3.9 from 3.5. We could not move to 3.10 nor 3.11 due to other issues we were facing with those releases (related to networking and recovering from split brain).
We cannot justify retesting with Hazelcast 3.12 at this point unfortunately. 
Our deployments are fairly static and by the time we roll-out new version with 3.12  there will be newer release of Hazelcast anyway.

Thanks for your reply.
Regards
Daniel
